src/libutil/util.cc

```diff
@@ -786,13 +786,64 @@ pid_t Pid::release()
     return p;
 }
 
-
 void killUser(uid_t uid)
 {
     debug(format("killing all processes running under uid '%1%'") % uid);
 
     assert(uid != 0); /* just to be safe... */
 
+#ifdef __APPLE__
+    /* We originally had very similar logic here as on the Linux side, but
+       then got bitten hard by a system crashing bug as described in here:
+       https://github.com/NixOS/nix/issues/1583. So now we adopt this other
+       approach, but we should go back to the kill approach once Apple fixes
+       their kernel. */
+
+    // Borrowed from https://stackoverflow.com/a/7733928
+    int err = 0;
+    size_t length = 0, newlength = 0;
+
+    // XXX: Should this be KERN_PROC_RUID? EUID vs. real UID
+    static const int name[4] = { CTL_KERN, KERN_PROC, KERN_PROC_UID, static_cast<int>(uid) };
+
+    int tries = 100;
+    bool done = false;
+
+    while (!done) {
+        tries--;
+
+        // First call with NULL to get number of processes. Yeah yeah TOCTOU is fun, but we check
+        // that it hasn't changed later and keep going until it's 0
+        err = sysctl((int *)name, 4, NULL, &length, NULL, 0);
```

Comment:
Does this work for you @copumpkin ?

Trying this little program is giving really big length values:

```
#include <sys/sysctl.h>
#include <stdio.h>
#include <string.h>

#define UID 30001

int main() {
        size_t length = 0;
        int errno = 0;
        static const int name[4] = { CTL_KERN, KERN_PROC, KERN_PROC_UID, UID };
        errno = sysctl((int *)name, 4, NULL, &length, NULL, 0);
        printf("uid:%i length:%zu err:%s\n", UID, length, strerror(errno));
}
```

Reply:
Maybe I screwed it up! I haven't actually run this yet 😄

Reply:
Haha that's fine! I'll go through it and see whats works. Before now I had hoped apple would just fix the issue but 10.13.1 still has the issue.

Reply:
Well each process entry is actually a struct so that might be fairly large. How big are the values you're seeing? Also, now that I think about it, my vector might be too big, since I should probably divide it by the size of the struct.

Reply:
Yep that looks more like it (ceacd3e565931797e3e4447a2649c78dfc3b3d45):

Although it looks like it's the same regardless of the user:

```
euid:501 uid:501 length:344 err:Undefined error: 0
euid:501 uid:30001 length:344 err:Undefined error: 0
```

Reply:
Weird, maybe we'll just need to filter by hand based on the struct?

Reply:
Or maybe try `KERN_PROC_RUID`